rfcs/0050-vm-syscalls-3/0050-vm-syscalls-3.md

```diff
@@ -0,0 +1,214 @@
+---
+Number: "0050"
+Category: Standards Track
+Status: Draft
+Author: Xu Jiandong <lynndon@gmail.com>, Dingwei Zhang <zhangsoledad@gmail.com>
+Created: 2023-04-17
+---
+
+# VM Syscalls 3
+
+## Abstract
+
+This document describes the addition of the syscalls during the CKB2023. This update significantly enhances the flexibility of CKB Script.
+
+The following four syscalls are added:
+
+- [Spawn]
+- [Get Memory Limit]
+- [Current Memory]
+- [Set Content]
+- [Load Block Extension]
+
+### Spawn
+[Spawn]: #spawn
+
+The syscall Spawn is the core part of this update. The *Spawn* and the latter two syscalls: *Get Memory Limit* and *Set Content* together, implement a way to call another CKB Script in a CKB Script. Unlike the *Exec*[1](../0034-vm-syscalls-2/0034-vm-syscalls-2.md) syscall, *Spawn* saves the execution context of the current script, like [posix_spawn](https://man7.org/linux/man-pages/man3/posix_spawn.3.html), the parent script blocks until the child script ends.
+
+
+```c
+typedef struct spawn_args_t {
+  uint64_t memory_limit;
+  int8_t* exit_code;
+  uint8_t* content;
+  uint64_t* content_length;
+} spawn_args_t;
+
+int ckb_spawn(size_t index, size_t source, size_t bounds,
+              int argc, char* argv[], spawn_args_t* spgs);
+```
+
+The arguments used here are:
+
+- `index`: an index value denoting the index of entries to read.
+- `source`: a flag denoting the source of cells to locate, possible values include:
+    - 1: input cells.
+    - `0x0100000000000001`: input cells with the same running script as current script
+    - 2: output cells.
+    - `0x0100000000000002`: output cells with the same running script as current script
+    - 3: dep cells.
+- `bounds`: high 32 bits means `offset`, low 32 bits means `length`. If `length` equals to zero, it read to end instead of reading 0 bytes.
+- `argc`: argc contains the number of arguments passed to the program
+- `argv`: argv is a one-dimensional array of strings
+- `memory_limit`: an integer value denoting the memory size to use(Not including descendant children scripts), possible values include:
+    - 1 (0.5 M)
+    - 2 (1 M)
+    - 3 (1.5 M)
+    - 4 (2 M)
+    - 5 (2.5 M)
+    - 6 (3 M)
+    - 7 (3.5 M)
+    - 8 (4 M)
+- `exit_code`: an int8 pointer denoting where we save the exit code of a child script.
+- `content`: a pointer to a buffer in VM memory space denoting where we would load the sub-script data. The child script will write data in this buffer via `set_content`.
```

Comment:
What happens when passing the pointer 0 as content?

Reply:
If the subscript does not return data, then using 0 or NULL is correct; if the subscript returns data and 0 or NULL is passed in at this time, unexpected memory errors may occur.